kartothek/io/dask/_update.py

```diff
@@ -51,14 +87,16 @@ def _update_dask_partitions_shuffle(
     if ddf.npartitions == 0:
         return ddf
 
+    group_cols = partition_on[:]
     if num_buckets is not None:
         meta = ddf._meta
         meta[_KTK_HASH_BUCKET] = np.uint64(0)
         ddf = ddf.map_partitions(_hash_bucket, bucket_by, num_buckets, meta=meta)
-        group_cols = [partition_on[0], _KTK_HASH_BUCKET]
-    else:
-        group_cols = [partition_on[0]]
+        group_cols.append(_KTK_HASH_BUCKET)
```

Comment:
why this change in logic?

Reply:
unrelated. this was a ill informed "optimization" when I initially created this. It is, however, quite harmful and creates possible very large hotspots. Consider a partitioning like
`["year", "month", "day"]` which would data data of each year on a single worker even though the data could be chunked much more granular.

Generally speaking, the less this is restricted, the more homogeneous the intermediate jobs will distribute the data and we should allow as much freedom here as possible. in the end the load balancing is done via a hash bucketing more or less identical to how we calculate buckets -> more input values will yield a better distribution in buckets (statistically speaking, there are of course exceptions)

Reply:
Thanks for the explanation. I think this may be more appropriate in a different commit, do you agree?

Reply:
If anything, this could be a separate PR. Since this is internal and the user facing thing will be "Significant improvement" in performance I kept it the way it is